File: VB/Grid_RowDetailsTemplateSelector/RowDetailsTemplateSelector.vb
Imports DevExpress.Xpf.Grid
Imports System
Imports System.Linq
Imports System.Windows
Imports System.Windows.Controls

Namespace Grid_RowDetailsTemplateSelector
	Friend Class RowDetailsTemplateSelector
		Inherits DataTemplateSelector

		Public Property PrimaryRowDetailsTemplate() As DataTemplate
		Public Property SecondaryRowDetailsTemplate() As DataTemplate
		Public Overrides Function SelectTemplate(ByVal item As Object, ByVal container As DependencyObject) As DataTemplate
			Dim row As RowData = TryCast(item, RowData)
			If row IsNot Nothing Then
                Dim _person As Person = TryCast(row.Row, Person)
                If _person IsNot Nothing Then
                    Return If(_person.Address.Contains("London"), PrimaryRowDetailsTemplate, SecondaryRowDetailsTemplate)
                End If
			End If
			Return MyBase.SelectTemplate(item, container)
		End Function
	End Class
End Namespace

File: CS/Grid_RowDetailsTemplateSelector/RowDetailsTemplateSelector.cs
using DevExpress.Xpf.Grid;
using System;
using System.Linq;
using System.Windows;
using System.Windows.Controls;

namespace Grid_RowDetailsTemplateSelector
{
    class RowDetailsTemplateSelector : DataTemplateSelector
    {
        public DataTemplate PrimaryRowDetailsTemplate { get; set; }
        public DataTemplate SecondaryRowDetailsTemplate { get; set; }
        public override DataTemplate SelectTemplate(object item, DependencyObject container) {
            RowData row = item as RowData;
            if (row != null) {
                Person person = row.Row as Person;
                if (person != null)
                    return person.Address.Contains("London") ? PrimaryRowDetailsTemplate : SecondaryRowDetailsTemplate;
            }
            return base.SelectTemplate(item, container);
        }
    }
}
